src/GafferScene/Instancer.cpp

```diff
@@ -857,10 +1471,99 @@ void Instancer::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
 		static_cast<AtomicCompoundDataPlug *>( output )->setValue( result );
 		return;
 	}
+	else if( output == variationsPlug() )
+	{
+		// Compute the number of variations by accumulating massive lists of unique hashes from all EngineDatas
+		// and then counting the total number of uniques
+		tbb::spin_mutex locationMutex;
+		std::vector< std::unique_ptr< EngineData::PrototypeHashes > > perLocationHashes;
+
+		std::function<bool(const GafferScene::ScenePlug *, const GafferScene::ScenePlug::ScenePath &)> functor =[this, &locationMutex, &perLocationHashes]( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
+		{
+			ConstEngineDataPtr engine = boost::static_pointer_cast<const EngineData>( enginePlug()->getValue() );
+			std::unique_ptr< EngineData::PrototypeHashes > locationHashes = engine->uniquePrototypeHashes();
+
+			tbb::spin_mutex::scoped_lock lock( locationMutex );
+			perLocationHashes.push_back( std::move( locationHashes ) );
+			return true;
+		};
+
+		GafferScene::SceneAlgo::filteredParallelTraverse( inPlug(), filterPlug(), functor );
+
+		CompoundDataPtr result = new CompoundData;
+
+		std::vector< int > numUnique;
+		std::vector< InternedString > outputNames;
+		if( perLocationHashes.size() == 0 )
+		{
+			// There is always an entry for the empty string, which contains the total variations
+			// for all context variables combined
+			result->writable()[""] = new IntData( 0 );
+		}
+		else
+		{
+			if( perLocationHashes.size() == 1 )
+			{
+				// We only have one location, so we can just output the sizes of the hash sets
+				// we got
+				for( const auto &var : *perLocationHashes[0] )
+				{
+					result->writable()[var.first] = new IntData( var.second.size() );
+				}
+			}
+			else
+			{
+				// For multiple locations, we need to merge the hash sets into a single giant set,
+				// and then check its size.  This seems very expensive, but we only do this when
+				// users are using the Context Variation tab, and need a display of how many
+				// variations they are creating.  This plug isn't evaluated at render time.
+				EngineData::PrototypeHashes combine( *perLocationHashes[0] );
+				for( unsigned int i = 1; i < perLocationHashes.size(); i++ )
+				{
+					for( auto &var : *perLocationHashes[0] )
+					{
+						if( combine.find( var.first ) == combine.end() )
+						{
+							combine[ var.first ] = var.second;
+						}
+						else
+						{
+							combine[ var.first ].merge( var.second );
+						}
```

Comment:
Is `combine[var.first] = var.second` definitely faster than `merge()` when the container is empty? Feels like a single `combine[var.first].merge( var.second )` might deal with all cases sufficiently?